src/Upload/StreamableUploader.php

```diff
@@ -0,0 +1,165 @@
+     * then we will send a chunk. Any remaining data is sent on close.
+     *
+     * @param mixed $data The data being written. Can be a string or stream.
+     * @return int The number of bytes written.
+     */
+    public function write($data)
+    {
+        // append data onto buffer
+        $this->buffer->write($data);
+        $this->upload(false);
+        return strlen($data);
+    }
+
+    /**
+     * Triggers the upload process.
+     *
+     * @param bool $remainder If true, send the all the remaining data and close
+     *        the file. Otherwise, only write data if we have enough to send a
+     *        chucked set.
+     * @return array
+     * @throws GoogleException
+     */
+    public function upload($remainder = true)
+    {
+        // determine how much data to write
+        $writeSize = $this->getChunkedWriteSize($remainder);
+        if ($writeSize == 0) {
+            return [];
+        }
+
+        // find or create the resumeUri
+        $resumeUri = $this->getResumeUri();
+
+        $rangeStart = $this->rangeStart;
+        $rangeEnd = $remainder ? "*" : $rangeStart + $writeSize - 1;
+
+        $data = $this->buffer->read($writeSize);
+
+        // do the streaming write
+        $headers = [
+            'Content-Length'    => $writeSize,
+            'Content-Type'      => $this->contentType,
+            'Content-Range'     => "bytes $rangeStart-$rangeEnd/*"
+        ];
+        $request = new Request(
+            'PUT',
+            $resumeUri,
+            $headers,
+            $data
+        );
+
+        try {
+            $response = $this->requestWrapper->send($request, $this->requestOptions);
+        } catch (Google\Cloud\Exception\ServiceException $ex) {
+            throw new GoogleException(
+                "Upload failed. Please use this URI to resume your upload: $resumeUri",
+                $ex->getCode(),
+                $ex
+            );
+        }
+
+        // reset the buffer with the remaining contents
+        $this->resetBuffer($this->buffer->getContents());
+        $this->rangeStart += $writeSize;
+
+        return json_decode($response->getBody(), true);
+    }
+
+    /**
+     * Determines the length of content to write
+     *
+     * @return int
+     */
+    private function getChunkedWriteSize($remainder)
+    {
+        $bufferSize = $this->buffer->getSize();
+        if ($remainder) {
+            return $bufferSize;
+        } else {
+            return floor($bufferSize / $this->chunkSize) * $this->chunkSize;
+        }
+    }
+
+    /**
+     * Finish writing the rest of the file.
+     */
+    public function close()
+    {
+        $this->upload();
+    }
+
+    /**
+     * After we've sent data, create a new buffer with the remaining data.
+     *
+     * @param string $data The data to store.
+     */
+    private function resetBuffer($data = "")
+    {
+        $this->buffer = new BufferStream($this->chunkSize);
```

Comment:
If we do end up keeping this class, don't forget to declare the properties with their visibility.